src/Computation/Expression.cs

```diff
@@ -198,7 +199,7 @@ public IExpression Select(IExpression trueValue, IExpression falseValue)
                 ? v.AsBool()
                     ? trueValue
                     : falseValue
-                : Ternary(trueValue, falseValue, Values.Select.Create)
+                : Ternary(trueValue, falseValue, (p, t, f) => new Select(p, t, f))
```

Comment:
It's probably fairly marginal, but this way can eliminate unnecessary constraints being built up on an `Expression`. You are correct that this implementation could be just

    public IExpression Select(IExpression trueValue, IExpression falseValue)
    {
        return trueValue.Size == falseValue.Size
            ? Ternary(trueValue, falseValue, Values.Select.Create)
            : throw new InconsistentExpressionSizesException(trueValue.Size, falseValue.Size);
    }

and the value creation logic would be

    public static IValue Create(IValue predicate, IValue trueValue, IValue falseValue)
    {
        return predicate is IConstantValue p
            ? p.AsBool()
                ? trueValue
                : falseValue
            : new Select(predicate, trueValue, falseValue);
    }

which would be more consistent with how everything else is done. However, if the `predicate` value is constant this would unnecessarily go through the whole unpacking and repacking of an `Expression` just to return one of the true or false values. This in itself isn't a huge deal since it is just a couple of allocations. The meaningful difference is that both of those true and false values could have some associated constraints, and instead of eliminating the set that goes with the unchosen side of the `Select` you end up combining them regardless.

However, this raises a larger point that something similar is probably already going on with our read-write optimizations, and in fact constraints living on `Expresssion` is also already causing me some small issues with the further SMT work I am trying to do here. So I am starting to think that the constraints should probably live on the value to trivially kill a couple of classes of potential issue, despite the fact I think there will be some extra verbosity and some ugliness this way.

We should discuss this. Could be important and/or enlightening for some other things.

Reply:
OK, in which case consistency here doesn't matter as it sounds like we should be doing more of this kind of thing in other places and so I think things will probably eventually become consistent again in another, more optimised, way.